File: btit_94.rs
// Definition for a binary tree node.
// #[derive(Debug, PartialEq, Eq)]
// pub struct TreeNode {
//   pub val: i32,
//   pub left: Option<Rc<RefCell<TreeNode>>>,
//   pub right: Option<Rc<RefCell<TreeNode>>>,
// }
//
// impl TreeNode {
//   #[inline]
//   pub fn new(val: i32) -> Self {
//     TreeNode {
//       val,
//       left: None,
//       right: None
//     }
//   }
// }
use std::cell::RefCell;
use std::rc::Rc;

#[derive(Debug)]
struct TreeNode {
    val: i32,
    left: Option<Rc<RefCell<TreeNode>>>,
    right: Option<Rc<RefCell<TreeNode>>>,
}

impl TreeNode {
    fn new(val: i32) -> Self {
        TreeNode {
            val,
            left: None,
            right: None,
        }
    }
}

impl Solution {
    pub fn inorder_traversal(root: Option<Rc<RefCell<TreeNode>>>) -> Vec<i32> {
        let mut response = Vec::new();
        let mut stack = Vec::new();
        let mut now = root;

        while now.is_some() || !stack.is_empty() {
            while let Some(node) = now {
                stack.push(node.clone());
                now = node.borrow().left.clone();
            }

            if let Some(node) = stack.pop() {
                response.push(node.borrow().val);
                now = node.borrow().right.clone();
            }
        }

        return response;
    }
}

File: js/btit_94.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[]}
 */
var inorderTraversal = function (root) {
  const response = [];
  const stack = [];
  let now = root;
  while (root || stack) {
    while (root) {
      stack.push(root);
      root = root.left;
    }
    root = stack.pop();
    response.push(root.val);
    root = root.right;
  }
  return response;
};

// another approach
var inorderTraversal = function (root) {
  const response = [];
  traversal(root);
  return response;
};

var traversal = function (root, res) {
  if (!root) return;
  if (root.left) traversal(root.left, res);
  res.push(root.val);
  if (root.right) traversal(root.right, res);
};
